### PR TITLE
Graph coloring

### PR DESCRIPTION
**Pull request by niveshd**
Using ZOLTAN coloring algorithm, to color Bidirected graphs represented by SparsityPattern. 

A color_sparsity_pattern function is added which colors a graph such that directly connected nodes don't have same color. The query functions needed are already present (needed for implementation of partitioning algorithm using ZOLTAN).

### CODE REVIEW

File: include/deal.II/lac/sparsity_tools.h [hunk 1]
```diff
@@ -103,6 +103,33 @@ namespace SparsityTools
...
                  );
 
+  /**
+   * Using a coloring algorithm provided by ZOLTAN to color indices whose
+   * connections are represented using a SparsityPattern object. In effect,
+   * we view this sparsity pattern as a graph of connections between nodes,
+   * where each nonzero entry in the sparsity pattern corresponds to
+   * an edge between two nodes in the connection graph. The goal is to assign
+   * each node a color index such that no two connected nodes have
+   * same color index. Note that this function requires that SparsityPattern be
```
Comment (davydden) [inline include/deal.II/lac/sparsity_tools.h]: `require the @p sparsity_pattern to be symmetric`

File: include/deal.II/lac/sparsity_tools.h [hunk 2]
```diff
@@ -103,6 +103,33 @@ namespace SparsityTools
...
+   * symmetric (and hence square) i.e a bidirected graph representation. We do
+   * not check for symmetry of the sparsity pattern, since this is an
+   * expensive operation, but rather leave this as the responsibility of
+   * caller of this function.
+   *
+   * @image html color_bidirected_graph.png
+   * For example, the image shows 5 nodes numbered from 0 to 4. The connections
```
Comment (davydden) [inline include/deal.II/lac/sparsity_tools.h]: > the image above shows...

File: include/deal.II/lac/sparsity_tools.h [hunk 3]
```diff
@@ -103,6 +103,33 @@ namespace SparsityTools
...
+   * same color.
+   *
+   * If deal.II was not installed with package ZOLTAN, this function will
+   * generate an error.
+   *
+   * Zoltan coloring algorithm is run in serial by each processor. Hence all
+   * processors have coloring information of all the nodes.
+   */
+  void color_sparsity_pattern (const SparsityPattern     &sparsity_pattern,
+                               std::vector<int> &color_indices);
```
Comment (davydden) [inline include/deal.II/lac/sparsity_tools.h]: please use `std::vector<unsigned int>` as an output.

Also I think it would make sense to return the total number of colors. 
Of course, one can just loop through `color_indices` to get maximum value: `*std::max_element(color_indices.begin(), color_indices.end());` if Zoltan does not provide this info.

File: source/lac/sparsity_tools.cc [hunk 4]
```diff
@@ -183,7 +183,12 @@ namespace SparsityTools
       Assert ( numEdges != nullptr , ExcInternalError() );
 
       for (int i=0; i<num_obj; ++i)
-        numEdges[i] = graph->row_length(globalID[i])-1;
+        {
+          if ( graph->exists(i,i) )     //Check if diagonal element is present
```
Comment (davydden) [inline source/lac/sparsity_tools.cc]: so does the new function also work in case user have diagonal entries at the sparsity pattern? That should not matter for coloring, of course.

Maybe add additional test (a copy of the present one) with the only difference being diagonal entries in sparsity pattern.

File: source/lac/sparsity_tools.cc [hunk 5]
```diff
@@ -326,6 +331,66 @@ namespace SparsityTools
...
+  void color_sparsity_pattern (const SparsityPattern     &sparsity_pattern,
+                               std::vector<int> &color_indices)
+  {
+    // Make sure that ZOLTAN is actually
+    // installed and detected
+#ifndef DEAL_II_TRILINOS_WITH_ZOLTAN
+    (void)sparsity_pattern;
+    AssertThrow (false, ExcZOLTANNotInstalled());
+#else
+    //Zoltan coloring algorithm is run in series by each processor. All objects colored are assumed to belong to Processor 0.
```
Comment (davydden) [inline source/lac/sparsity_tools.cc]: > All objects colored are assumed to belong to Processor 0.

i don't think that's the case, you assume that exactly the same objects are there on each MPI process.

File: source/lac/sparsity_tools.cc [hunk 6]
```diff
@@ -326,6 +331,66 @@ namespace SparsityTools
...
+    //Set query functions required by coloring function
+    zz->Set_Num_Obj_Fn (get_number_of_objects, &graph);
+    zz->Set_Obj_List_Fn (get_object_list, &graph);
+    zz->Set_Num_Edges_Multi_Fn (get_num_edges_list, &graph);
+    zz->Set_Edge_List_Multi_Fn (get_edge_list, &graph);
+
+    //Variables needed by coloring function
+    int num_gid_entries=1;
+    const int num_objects = graph.n_rows();
+    ZOLTAN_ID_PTR global_ids = new ZOLTAN_ID_TYPE [num_objects];   //Pointer type fixed by Color function. Size needs to be preallocated.
```
Comment (davydden) [inline source/lac/sparsity_tools.cc]: can't you use `std::vector< ZOLTAN_ID_PTR> global_ids(num_objects);` and avoid manually deleting? Same for `color_exp`.

File: source/lac/sparsity_tools.cc [hunk 7]
```diff
@@ -326,6 +331,66 @@ namespace SparsityTools
...
+
+    //Check for error code in partitioner
+    Assert (rc == ZOLTAN_OK , ExcInternalError());
+
+    //Allocate and assign color
+    color_indices.resize (num_objects);
+    std::copy (color_exp, color_exp + num_objects, color_indices.begin());
+
+    //Cleaning up
+    delete global_ids;
```
Comment (davydden) [inline source/lac/sparsity_tools.cc]: let's avoid `delete` if possible.

**Comment by davydden** [comment]
@dealii/dealii just to put this PR into context: @niveshd is my student who will later use coloring to add extra functionality that simplifies using `FE_Enriched` PUM with multiple overlapping enrichment domains.

File: tests/zoltan/zoltan_02.cc [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  color_indices.resize(num_indices);
+  SparsityTools::color_sparsity_pattern (sp_graph, color_indices);
+
+  //color
+  deallog << "Coloring" << std::endl;
+  for (unsigned int i=0; i<num_indices; i++)
+    {
+      deallog << i << " " << color_indices[i] << std::endl;
+    }
+}
```
Comment (davydden) [inline tests/zoltan/zoltan_02.cc]: new line

File: tests/zoltan/zoltan_02.cc [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  graph.add(4,1);
+  graph.add(3,4);
+  graph.add(4,2);
+  graph.add(2,3);
+}
+
+
+int main (int argc, char **argv)
+{
+  initlog();
```
Comment (davydden) [inline tests/zoltan/zoltan_02.cc]: look at other tests with `mpi` and please extend this one to run with more than one MPI core. That's done by adding another output file `zoltan_02.mpirun=3.output` and adding a few things here. Look at `tests/lac/linear_operator_14.cc`, for example.

File: include/deal.II/lac/sparsity_tools.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/lac/sparsity_tools.h]: -> have THE same color index

File: include/deal.II/lac/sparsity_tools.h [hunk 10]
```diff
@@ -103,6 +103,33 @@ namespace SparsityTools
...
+   * each node a color index such that no two connected nodes have
+   * same color index. Note that this function requires that SparsityPattern be
+   * symmetric (and hence square) i.e a bidirected graph representation. We do
+   * not check for symmetry of the sparsity pattern, since this is an
+   * expensive operation, but rather leave this as the responsibility of
+   * caller of this function.
+   *
+   * @image html color_bidirected_graph.png
+   * For example, the image shows 5 nodes numbered from 0 to 4. The connections
+   * shown are bidirectional. After coloring, no two connected nodes have the
```
Comment (bangerth) [inline include/deal.II/lac/sparsity_tools.h]: The commonly used term, I believe, is "undirected", nor "bidirectional". You also use this in the figure itself (which I can't seem to comment on in github).

File: include/deal.II/lac/sparsity_tools.h [hunk 11]
```diff
@@ -103,6 +103,33 @@ namespace SparsityTools
...
+   * same color.
+   *
+   * If deal.II was not installed with package ZOLTAN, this function will
+   * generate an error.
+   *
+   * Zoltan coloring algorithm is run in serial by each processor. Hence all
```
Comment (bangerth) [inline include/deal.II/lac/sparsity_tools.h]: in lines 124 and 127, pick one way to spell Zoltan. Same everywhere.

File: source/lac/sparsity_tools.cc (hunk 5 above)
Comment (bangerth) [reply source/lac/sparsity_tools.cc]: Here and below, can you break long lines at approximately 80 characters?

**Comment by davydden** [comment]
@bangerth :

> There is already a function that does graph coloring in include/deal.II/base/graph_coloring.h. We should think about how they relate

maybe @Rombur or @kronbichler can comment, as they are the authors of the algorithm there.

My take is the following: Looking at this [make_graph_coloring](https://www.dealii.org/developer/doxygen/deal.II/namespaceGraphColoring.html) function, it takes graph input in a completely different way, also it looks like it's quite tailored to the assembly in WorkStream class as it takes functions which return `types::global_dof_index` to determine connectivity between nodes.

Bottom line, I don't think it's feasible to join the two. Would you find it acceptable if @niveshd mentions in the documentation something like: 

> @note The current function is a more general alternative to GraphColoring::make_graph_coloring() which is tailored to graph coloring arising in shared-memory parallel assembly of matrices.

**Comment by bangerth** [comment]
Yes, sure, something like this is ok. I don't think either of the two is more general -- it would be easy to convert from one way of providing input to the other. But linking the documentation of the two is just fine with me.

**Comment by davydden** [comment]
> I don't think either of the two is more general

what i meant is the function in this PR is not tailored to matrix assembly and specific representation of connectivity there.

**Comment by Rombur** [comment]
> what i meant is the function in this PR is not tailored to matrix assembly and specific representation of connectivity there.

Neither is `make_graph_coloring`.  `get_conflict_indices` can be recast as a sparsity pattern or you can write a function that give conflict indices from a sparsity pattern. It is just two different ways to represent the same information.

**Comment by davydden** [comment]
> get_conflict_indices can be recast as a sparsity pattern or you can write a function that give conflict indices from a sparsity pattern. 

sure, but the argument `const std::function< std::vector< types::global_dof_index >(const typename identity< Iterator >::type &)> &` clearly indicates that this function actually has to do with `types::global_dof_index`, which are specific to the problems addressed in WorkStream, don't you agree? Graph coloring per-se does not know what ` std::vector< types::global_dof_index >` is and does not have to.

**Comment by bangerth** [comment]
The `std::vector` is essentially the nonzero entries in the sparsity pattern of one row. It's easy to convert from one format to the other -- in fact, I bet it could be done with just a few lines of lambda functions if one wanted to.

But that's besides the point. I think this function is of independent interest!

**Comment by davydden** [comment]
> The std::vector is essentially the nonzero entries in the sparsity pattern of one row.

sure.

> It's easy to convert from one format to the other 

I am not arguing this. My point is that the existing function takes arguments in a way convenient for usage in WorkStream and matrix assembly. Otherwise it is kind of hard to motivate why graph coloring should have anything to do with `types::global_dof_index` and have different input when deal.II is compiled with 32bit integers and 64bit integers.

**Comment by Rombur** [comment]
> I am not arguing this. My point is that the existing function takes arguments in a way convenient for usage in WorkStream and matrix assembly.

I don't understand your point. We agree that both inputs give the same information. So basically, your problem is that we picked `types::global_dof_index` instead of another type? I could argue that the fact that you are using a `SparsityPattern` will limit you to color matrices.  Internally, `make_graph_coloring` is based on DSATUR which is a general coloring algorithm and we agree that the inputs are equivalent so I don't understand why one function is more general than the other.

**Comment by davydden** [comment]
@Rombur 

> We agree that both inputs give the same information

sure

> So basically, your problem is that we picked types::global_dof_index instead of another type?

it's not a problem. This choice indicate that the function is tailored to WorkStream. Same applies to taking iterators. If we take nodes as iterators `const Iterator & begin` and do not know anything about WorkStream and matrices, then I am (as a naïve user) expect connectivity to be represented with similar objects, namely one `Iterator` maps to a couple of other `Iterator` to represent edges in graph.  

I am **not** saying the function you came up with for WorkStream is bad, I am just saying it's taylored to the task. Which is reflected in arguments. If you distance yourself from WorkStream/matrices, then 
there is no reason to use `types::global_dof_index` as part of the arguments in graph coloring.

> I could argue that the fact that you are using a SparsityPattern will limit you to color matrices.

Absolutely not. As a matter a fact the unit test shows how to partition just the graph, without any matrices whatsoever. The usage of `SparsityPattern` is just because this class is (ab)used to partition graph in shared tria etc.

> I don't understand why one function is more general than the other.

because this function takes a graph in what `dealii` uses in other public functions to represent a graph -- `SparsityPattern`, whereas another function is tailored to be used with `WorkStream` and DoFs. The current function does not know anything about `types::global_dof_index`.

**Comment by bangerth** [comment]
I think we're all in agreement :-) Let's go back to focusing on the content of this patch.

@niveshd -- can you let us know when you have done the changes already requested?

**Comment by niveshd** [comment]
@bangerth @davydden i will let you know, when I have added the changes :) 
 Thank you all for your comments 👍

File: source/lac/sparsity_tools.cc (hunk 4 above)
Comment (niveshd) [reply source/lac/sparsity_tools.cc]: Yes, the new function works even if we have diagonal entries. In fact, the sparsity pattern already has diagonal element by default even if I didn't add diagonal elements to dynamic sparsity pattern (before copying it back to sparsity pattern).

File: include/deal.II/lac/sparsity_tools.h [hunk 12]
```diff
@@ -103,6 +104,40 @@ namespace SparsityTools
...
+   * nodes have the same color.
+   *
+   * If deal.II was not installed with package ZOLTAN, this function will
+   * generate an error.
+   *
+   * @note The current function is a more general alternative to
+   * GraphColoring::make_graph_coloring() which is tailored to graph
+   * coloring arising in shared-memory parallel assembly of matrices.
+   */
+  void color_sparsity_pattern (const SparsityPattern     &sparsity_pattern,
```
Comment (davydden) [inline include/deal.II/lac/sparsity_tools.h]: please, just return the number of colors, i.e. 

```
unsigned int color_sparsity_pattern (const SparsityPattern     &sparsity_pattern, std::vector<unsigned int> &color_indices)
```

File: tests/zoltan/zoltan_02.mpirun=2.output [hunk 13]
```diff
@@ -0,0 +1,8 @@
+
+DEAL::Coloring
+DEAL::Number of colors used: 4
```
Comment (davydden) [inline tests/zoltan/zoltan_02.mpirun=2.output]: it's not enough just to add those extra output files, you need to use in tests
```
Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
MPILogInitAll all;
```
and remove `initlog();`

Comment (davydden) [reply tests/zoltan/zoltan_02.mpirun=2.output]: otherwise you don't really test the output from each core.

File: tests/zoltan/zoltan_02.cc (hunk 8 above)
Comment (niveshd) [reply tests/zoltan/zoltan_02.cc]: I added a line after. Did you mean before 88

Comment (davydden) [reply tests/zoltan/zoltan_02.cc]: it's good the way it is currently. thanks.

File: tests/zoltan/zoltan_02.mpirun=2.output (hunk 13 above)
Comment (davydden) [reply tests/zoltan/zoltan_02.mpirun=2.output]: ps. `MPILogInitAll all;` is a helper class for tests which in constructor creates output file for each MPI core and make `deallog` use it, and on destruction joins them together into a single output file.

File: source/lac/sparsity_tools.cc (hunk 4 above)
Comment (niveshd) [reply source/lac/sparsity_tools.cc]: Should I still add an additional test with no diagonal elements? Currently, sparsity pattern automatically adds diagonal elements when copied from dynamic sparsity pattern. If the test is needed i have to find some other way.

Comment (davydden) [reply source/lac/sparsity_tools.cc]: > Should I still add an additional test with no diagonal elements?

no, i think it's fine.

File: include/deal.II/lac/sparsity_tools.h [hunk 14]
```diff
@@ -103,6 +104,38 @@ namespace SparsityTools
...
+   *
+   * @image html color_undirected_graph.png
+   * For example, the above image shows 5 nodes numbered from 0 to 4.
+   * The connections shown are bidirectional. After coloring, no two connected
+   * nodes have the same color.
+   *
+   * If deal.II was not installed with package ZOLTAN, this function will
+   * generate an error.
+   *
+   * @note The current function is a more general alternative to
```
Comment (davydden) [inline include/deal.II/lac/sparsity_tools.h]: please write `is an alternative to...`

**Comment by davydden** [comment]
@niveshd meanwhile you can squash it into two commits: first to update `zoltan_01` test, and the second for the new functionality.

**Comment by davydden** [comment]
@niveshd reset to c9f9d5f and then just squash commits. You last push was not right.

**Comment by niveshd** [comment]
I reset it to the commit you mentioned. Now I can't push because my remote branch is head. Should I perhaps use push -f.

I was trying to squash some commits. After squashing, I had to pull and merge. I think this caused the problem.

**Comment by davydden** [comment]
p.s. the easiest way for your to squash into two commits is
```
git reste --soft master
git add <-- here goes updates to the test
git commit -m "...."
git add <--- here comes everything else
git commit -m "..."
git diff <---you remote branch, should give you zero difference
git push -f
```

**Comment by Rombur** [comment]
> But then we have graph partitioning in SparsityTools:: and graph colouring in GraphColoring::. That is also a bit confusing. I would vote to keep the function in SparsityTools:: namespace.

I don't have a strong preference. You could also have a dummy function in `GraphColoring` which only calls the the function in `SparsityTools`. This way it is easy for someone to find out which `GraphColoring` algorithm we support and you can keep the implementation in `SparsityTools` since all the functions you need are there.

**Comment by bangerth** [comment]
>     But then we have graph partitioning in SparsityTools:: and graph colouring
>     in GraphColoring::. That is also a bit confusing. I would vote to keep the
>     function in SparsityTools:: namespace.
>
> I don't have a strong preference. You could also have a dummy function in |GraphColoring| which only calls the the function in |SparsityTools|. This way it is easy for someone to find out which |GraphColoring| algorithm we support and you can keep the implementation in |SparsityTools| since all the functions you need are there.

I also don't have a strong preference. @rombur's suggestion seems sensible to me.

**Comment by niveshd** [comment]
@davydden @bangerth i did the required changes. Please let me know if there is anything else.

File: source/lac/sparsity_tools.cc [hunk 15]
```diff
@@ -326,6 +331,68 @@ namespace SparsityTools
...
+    int rc = zz->Color (num_gid_entries,
+                        num_objects,
+                        global_ids.data(),
+                        color_exp.data());
+
+    //Check for error code
+    Assert (rc == ZOLTAN_OK , ExcInternalError());
+
+    //Allocate and assign color
+    std::copy (color_exp.begin(), color_exp.end(), color_indices.begin());
```
Comment (bangerth) [inline source/lac/sparsity_tools.cc]: I think you should make sure that `color_indices` has the right size before you execute this line.

Comment (davydden) [reply source/lac/sparsity_tools.cc]: @niveshd you can use 
```
Assert (color_indices.size() == <some-other-size>, 
        ExcDimensionMismatch(color_indices.size(), <some-other-size>);
```

**Comment by davydden** [comment]
@bangerth any further comments to Zoltan PR?
